### PR TITLE
join-methods for 'zoomed_dm'

### PR DESCRIPTION
**Pull request by TSchiefer**
- methods for `left_join()`, `inner_join()`, `full_join()`, `semi_join()`, `anti_join()`, `right_join()` for class `zoomed_dm`
-  if not explicitly given, `by` makes use of FK-relation
- argument `select` implemented as suggested in #92 
- tracking key columns even if renamed with suffix during join

### CODE REVIEW

**Comment by TSchiefer** [comment]
R CMD check fail comes from vignette-building fails (repaired by f3611a9 and 6d39cd9)

File: R/dplyr.R [hunk 1]
```diff
@@ -122,3 +122,123 @@ rename.zoomed_dm <- function(.data, ...) {
...
+left_join.dm <- function(x, ...) {
+  abort_no_table_zoomed_dplyr("left_join")
+}
+
+#' @export
+left_join.zoomed_dm <- function(x, y, by = NULL, copy = FALSE, suffix = c(".x", ".y"), select = NULL, ...) {
+  if (nrow(cdm_get_filter(x) %>% filter(table == !!orig_name_zoomed(x)))) abort_no_filters_rename_select()
+  y_name <- as_string(enexpr(y))
+  join_data <- prepare_join(x, y_name, by, enexpr(select), suffix[1])
+  if (copy) message("Tables in a `dm` are necessarily on the same `src`, setting `copy = FALSE`.")
```
Comment (krlmlr) [inline R/dplyr.R]: Can we move this into `prepare_join()` ?

File: R/dplyr.R [hunk 2]
```diff
@@ -122,3 +122,123 @@ rename.zoomed_dm <- function(.data, ...) {
...
+#' @export
+left_join.dm <- function(x, ...) {
+  abort_no_table_zoomed_dplyr("left_join")
+}
+
+#' @export
+left_join.zoomed_dm <- function(x, y, by = NULL, copy = FALSE, suffix = c(".x", ".y"), select = NULL, ...) {
+  if (nrow(cdm_get_filter(x) %>% filter(table == !!orig_name_zoomed(x)))) abort_no_filters_rename_select()
+  y_name <- as_string(enexpr(y))
+  join_data <- prepare_join(x, y_name, by, enexpr(select), suffix[1])
```
Comment (krlmlr) [inline R/dplyr.R]: Ideally we'd get rid of the line above and do

```suggestion
  join_data <- prepare_join(x, {{ y }}, by, {{ select }}, suffix[1])
```

with suitable `ensym()` and `enquo()` calls in `prepare_join()` .

File: R/dplyr.R [hunk 3]
```diff
@@ -122,3 +122,123 @@ rename.zoomed_dm <- function(.data, ...) {
...
+  if (copy) message("Tables in a `dm` are necessarily on the same `src`, setting `copy = FALSE`.")
+  joined_tbl <- inner_join(join_data$x_tbl, join_data$y_tbl, join_data$by, copy = FALSE, suffix = suffix, ...)
+  replace_zoomed_tbl(x, joined_tbl, join_data$new_key_names)
+}
+
+#' @export
+full_join.dm <- function(x, ...) {
+  abort_no_table_zoomed_dplyr("inner_join")
```
[the remark this answers is not in the document]
Comment (TSchiefer) [reply R/dplyr.R]: good point, using `check_zoomed()` now

**Comment by krlmlr** [comment]
Thanks! Let's get rid of the suffix in #147.

**Pull request merged**